sbncode/TPCReco/VertexStub/TwoPlaneStubMerge_tool.cc

```diff
@@ -0,0 +1,220 @@
+/**
+ *
+ */
+
+// Framework Includes
+#include "art/Framework/Principal/Event.h"
+#include "art/Framework/Principal/Handle.h"
+#include "art/Framework/Services/Registry/ServiceHandle.h"
+#include "art/Persistency/Common/PtrMaker.h"
+#include "art/Utilities/ToolMacros.h"
+#include "cetlib/cpu_timer.h"
+#include "fhiclcpp/ParameterSet.h"
+#include "messagefacility/MessageLogger/MessageLogger.h"
+
+// local includes
+#include "sbncode/TPCReco/VertexStub/IStubMerge.h"
+#include "sbncode/TPCReco/VertexStub/StubMergeAlgorithms.h"
+
+// LArSoft includes
+#include "larcorealg/Geometry/BoxBoundedGeo.h"
+#include "larcore/Geometry/Geometry.h"
+#include "larcorealg/Geometry/GeometryCore.h"
+#include "larcore/CoreUtils/ServiceUtil.h"
+
+// std includes
+#include <string>
+#include <iostream>
+#include <memory>
+
+namespace sbn{
+/**
+ *  @brief  TwoPlaneStubMerge class definiton
+ */
+class TwoPlaneStubMerge : public sbn::IStubMerge {
+public:
+    /**
+     *  @brief  Constructor
+     */
+    TwoPlaneStubMerge(fhicl::ParameterSet const &pset);
+
+    /**
+     *  @brief  Destructor
+     */
+    ~TwoPlaneStubMerge();
+
+    std::vector<sbn::StubInfo> Merge(const std::vector<sbn::StubInfo> &stubs,
+       const geo::GeometryCore *geo, 
+       const spacecharge::SpaceCharge *sce, 
+       const detinfo::DetectorClocksData &dclock,
+       const detinfo::DetectorPropertiesData &dprop) override;
+
+    sbn::StubInfo MergeStubs(const sbn::StubInfo &A, const sbn::StubInfo &B,
+      const geo::GeometryCore *geo,
+      const spacecharge::SpaceCharge *sce,
+      const detinfo::DetectorPropertiesData &dprop);
+
+    bool SortStubs(const sbn::StubInfo &A, const sbn::StubInfo &B);
+
+private:
+  sbn::PlaneTransform fPlaneTransform;
+  double fMaxMergeTOff;
+  double fMaxMergeQOff;
+  bool fRemoveDuplicateMerges;
+  bool fSaveOldStubs;
+};
+
+TwoPlaneStubMerge::TwoPlaneStubMerge(fhicl::ParameterSet const &pset):
+  fPlaneTransform(pset.get<fhicl::ParameterSet>("PlaneTransform")),
+  fMaxMergeTOff(pset.get<double>("MaxMergeTOff")),
+  fMaxMergeQOff(pset.get<double>("MaxMergeQOff")),
+  fRemoveDuplicateMerges(pset.get<bool>("RemoveDuplicateMerges")),
+  fSaveOldStubs(pset.get<bool>("SaveOldStubs"))
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+TwoPlaneStubMerge::~TwoPlaneStubMerge()
+{
+}
+    
+// Return true if stub A is "better" than stub B
+bool TwoPlaneStubMerge::SortStubs(const sbn::StubInfo &A, const sbn::StubInfo &B) {
+  // Count the hits on the main stub
+  int A_nhit = A.stub.CoreNHit();
+  int B_nhit = B.stub.CoreNHit();
+
+  // First -- try more hits
+  if (A_nhit != B_nhit) return A_nhit > B_nhit;
+
+  // Sort by desirability of plane: 2 better than 0 better than 1
+  return ((A.vhit_hit->WireID().Plane + 1) % 3) < ((B.vhit_hit->WireID().Plane + 1) % 3);
```

Comment:
Do we want to rely on the numbering scheme here, or to check for 'collection vs not? or to make it configurable what plans to set as desirable? I'm just not sure if 0 vs. 1 holds up over time.

Reply:
The 1st induction plane is also better than the second. Is there a way to express that without relying on the Plane number?

Reply:
Two other notes: this piece of code isn't making any analysis decision -- it just picks which stub goes first in the output vector. So it does not matter if the ordering doesn't hold up over time.

Also, I don't see any reason to expect that the 2nd induction would perform better than the first in ICARUS, SBND, or uBooNE.

Reply:
Well, but I thought that the output vector was ordered such that the "best" plane was the first in the list? Or I am likely misunderstanding something...maybe that was a different output vector in the Stub?

But, that's fair enough. It would be good if this was documented. Is there a presentation or something that we can reference for the future?

Reply:
That's correct, it tries to place the best plane first. But it doesn't need to do that perfectly.

This is the most recent presentation on what this code does: https://sbn-docdb.fnal.gov/cgi-bin/private/ShowDocument?docid=21315.